File: crm_integration/public/js/work_order.js
frappe.ui.form.on("Work Order", {
	setup: function(frm) {
		set_sales_order_query(frm);
	},

	onload: function(frm) {
		set_sales_order_query(frm);
	},

	production_item: function(frm) {
		set_sales_order_query(frm);
	}
});

function set_sales_order_query(frm) {
	frm.set_query("sales_order", function() {
		if (frm.doc.production_item) {
			return {
				query: "crm_integration.crm_integration.work_order.query_sales_order",
				filters: {
					production_item: frm.doc.production_item,
					company: frm.doc.company
				}
			};
		}

		return {
			filters: {
				docstatus: 1,
				status: ["not in", ["Closed", "On Hold"]]
			}
		};
	});
}

File: crm_integration/crm_integration/work_order.py
import frappe
@frappe.whitelist()
@frappe.validate_and_sanitize_search_inputs
def query_sales_order(doctype, txt, searchfield, start, page_len, filters):
	"""Return submitted Sales Orders available for Work Order creation."""
	filters = frappe._dict(filters or {})
	production_item = filters.get("production_item")
	company = filters.get("company")

	if not production_item:
		return []

	base_filters = [["Sales Order", "docstatus", "=", 1]]
	if company:
		base_filters.append(["Sales Order", "company", "=", company])
	base_filters.append(["Sales Order", "status", "not in", ["Closed", "On Hold"]])

	return frappe.get_list(
		"Sales Order",
		fields=["name"],
		filters=base_filters,
		or_filters=[
			["Sales Order Item", "item_code", "=", production_item],
			["Packed Item", "item_code", "=", production_item],
		],
		as_list=True,
		distinct=True,
	)

